### ISSUE TITLE
(작성중)Cross-modal Prototype Driven Network for Radiology Report Generation

### ISSUE THREAD

**Issue by iamhxxn2** [iRAIL-SNU/irail-paper-review#16]
# Motivation

- **Radiology report generation(RRG)의 목적은** radiology image를 human-like language를 묘사하는 것이고, 이러한 과정은 manual하게 reporting하는 radiologists의 부담을 줄여준다.

- 이와 관련된 사전 연구들 중 몇몇의 연구에서는 cross-modal feature interaction에 대한 연구가 이루어졌지만, 대부분의 연구에서는 **encoder-decoder architecture**를 적용하고, **single-modal feature learning**에 집중하였다. 이러한 방법론들은 **complex cross-modal patterns을 capture하는데 어려움 존재**한다. 

- 본 연구에서는 radiology report generation task을 개선시키고, cross-modal pattern learning을 위해 **Cross-modal PROtotype driven NETwork (XPRONET)을 제안**한다.

- **Cross-modal PROtotype driven NETwork (XPRONET)의 세 가지 modules**
    -  **shared cross-modal prototype matrix** to record the cross modal prototypes
    -  **cross-modal prototype network** to learn the corss-modal prototypes and embed the corss-modal information into the visual and textual features
    - **improved multi-label contrastive loss**

# Method

## XPRONET의 **세 가지 main modules**
   1. **image feature extractor**
   2. **cross-modal prototype network**
   3. **encoder-decoder**  

![image](https://user-images.githubusercontent.com/97836929/220656571-be1ce345-cf08-48c4-a8fb-bd068a8cf899.png)
- **Figure : The architecture of XPRONET**
   - **time step T**에서의 word(e.g. "lungs")가 **embedding layer를 통해서 embedding**된다.
   - **visual & textual representations이 pseudo label을 기반으로 선택된 cross-modal prototypes과 cross-modal interaction을 수행**하기 위해 **cross-modal prototype querying and responding module로 이동**한다.
   -  report를 생성하기 위해 **linear layer를 통해서 reponses과 single-modal features이  각각 Transformer encoder-decoder의 input으로 사용**된다. 

### 1. image Feature Extractor
   - Model : **pre-trained ResNet-101** 
   - radiology image, $I$가 들어갔을 때, **image feature $v \in R^{H \times C \times W}$** 
   - **H, W, C**는 각각 **이미지의 height, width, Channel의 수**를 의미
   - ##ㅇㄹㅇㄹ

#### 2.1 Pseudo Label Generation 
- Cross-modal prototype은 각각의 sample에 대한 category 정보가 필요하다. 
- CheXbert를 이용하여 각각의 image-text pair에 대한 pseudo label을 생성
- $R={w_1, w_2, ..., w_{N^r-1}, w_{N^r}}$
   - $w_i$ : report에 존재하는 $i$번째 word
   - $N^r$ : report에 존재하는 words의 수
- ${y_1, y_2, ..., y_{N^l-1}, w_{N^l}}=f_{al}(R)$
   - one-hot vector로 표현
   - $N^l$ : category의 수
   - $f_{al}()$ : automatic radiology report labeler
   
#### 2.2 Pseudo Matrix Initialization
 - cross-modal pattern learning을 잘 수행하기 위해, **shared cross-modal prototype matrix** design
- shared cross-modal prototype matrix
   - **cross-modal patterns을 학습하고, 저장**하는 역할 
   - training 과정에서 update
   - **class-related prototype querying and responding modules에서 이용**된다.
- shared cross-modal prototype matrix을 random하게 initialize하지 않고, **prior information을 이용**한다.
   1. **pseudo class labels $y$를 가지는 image-text pair**에 대해서, **pre-trained ResNet-101과 BERT**를 이용하여 **global visual & textual representations 추출** 
      - $o^i \in R^{1 \times C_1}$ 
         $o^t \in R^{1 \times C_2}$
         $C_1, C_2$은 각각 추출된 visual, textual representations의 channel 수
      - robustness를 위해 flipped image features을 추출 $o^{if} \in R^{1 \times C_1}$
   2. 모든 training samples에 대해서 위의 process를 추출 
**$R^I_k$ = { $o_u^{i(f)}| y_{u,k}$ =1}**
**$R^T_k$ = { $o_u^t| y_{u,k}$ =1}**
$R^I_k$ , $R^T_k$는 각각 category k에 속하는 **visual, textual feature sets**을 의미
   3. cross-modal features을 형성하기 위해 **visual, textual representations을 concatenation** 
    **$r \in R^{1 \times D }$, $D = C_1 + C_2$**
   4. **K-Means**을 적용하여 각각의 **feature set을 $N_p$개의 clusters로 clustering 진행**
   5. 최종적으로 **각각의 cluster에 있는 features의 평균값**이 **PM의 initial cross-modal prototype으로 이용**이 된다.
- 위의 process에 대한 수식
![image](https://user-images.githubusercontent.com/97836929/220853373-74347ad3-c41f-436f-9bc4-2e6bd6798d3f.png)
        $o_u$ : sample u에 대한 concatenated cross-modal representation
        $R_k$ : category k에 속하는 cross-modal feature set
        $g_i^k$ : $k^{th}$에 대해 $i^{th}$ grouped cluster
        $N_{k,i}^d$ : $k^{th}$에 대해 $i^{th}$ grouped cluster에 존재하는 samples의 수
        $PM(k, i)$ : $k^{th}$에 대해 cross-modal prototype set에 있는 $i^{th}$ vector

#### 2.3 Cross-modal Prototype Querying
- 본 논문에서는 **cross-modal information을 single-modal features로 embeding** 시키기 위해 **Querying and reponding process를 적용**
- **Process**
   1. image-text training pair <I, R>과 pseudo label y이 주어지면, **각 sample에 대한 quried cross-modal prototype vectors(pv)생성** 한다.   
![image](https://user-images.githubusercontent.com/97836929/220861285-594b75cd-4eb7-428e-be7d-f03d155846ce.png)
PM(k) : 위에서 생성된  k^{th} category 에 대한 cross-modal prototype set을 의미

   2.  noise를 filtering하기 위해, **pv vectors을 $C_p$ dimension으로 mapping 시키는 linear projection 진행**
![image](https://user-images.githubusercontent.com/97836929/220862671-843819a9-ebd4-4504-ab2d-96bd6e04c40d.png)
$W_{pv} \in R^{D \times C_p }$

   3. 이미지와 동일한 label에 대해 **single-modal representation과 cross-modal prototype vectors 사이 similarity**
      - embedding layer를 거친 report representation output 
     : $v_t =$ { $v^t_1, v^t_2,..., v^t_{N^t-1}, v^{N^t}$ }, $v_i^t \in R^{1 \times C }$은 report에 있는 $i^{th}$ word embedding
     cross-prototype vector : $p_i$
      - querying을 진행하기 전, visual feature sequence $v_s$, textual report embeddings $v_t$, cross-modal prototype vector $p_i$의 dimension이 모두 다르기 때문에 dimension d로 동일하게 설정
![image](https://user-images.githubusercontent.com/97836929/220866866-6dc2cc03-b271-4ba3-8cd7-a6fc315111e5.png)
$W_{v} \in R^{C \times d }$, $W_{p} \in R^{C_p \times d }$
      - 다음 식과 같이, single-modal feature과 cross-modal prototype vector pair 사이의 similarity 계산
![image](https://user-images.githubusercontent.com/97836929/220867640-ea744bc7-961a-4ad5-9141-76f302370c0f.png)

   4. single-modal representation과 interact하기 위해 **가장 similarity가 높은 $\gamma$개의 vectors 선택**
      - cross-modal prototypes 중 대부분이 quried vector와 연관성이 떨어지기 때문에 $\gamma$개의 vectors 선택
   
   5. similarities를 기반으로 선택된 prototype vectors에 대해서 weights 계산
![image](https://user-images.githubusercontent.com/97836929/220914244-d969c126-ce99-429f-9fff-127ac468c608.png)

#### 2.4 Cross-modal Prototype Responding
- 상위 $\gamma$개의 similar cross-modal prototype vectors과 weights를 구한 이후 **다음 단계로 visual, textual features에 대한 responses 생성**
- **Process**
   1. **fully connected layer**를 통해서 queried prototype vectors을 **query vectors의 동일한 representation space로 변환**
![image](https://user-images.githubusercontent.com/97836929/220920081-01637661-ecdd-441e-84a7-5d757cd6b54c.png)
   2. **$p_{(i,j)}^{s \star}$ , $p_{(i,j)}^{t \star}$** : 각각 **$i^{th}$ image patch and word에 대한 cross-modal prototype**에 있는 **$j^{th}$ prototype**을 의미
![image](https://user-images.githubusercontent.com/97836929/220920181-fe7f56fe-4a81-40df-98d3-fe57970b21f1.png)
      - **$e_{(i,j)}^{s}$ , $e_{(i,j)}^{t}$** : $i^{th}$ image patch와 word에 대한 **$j^{th}$ transformed prototype vectors**
      -  **$r_{i}^{s}$ , $r_{i}^{t}$** : **$i^{th}$ image patch와 word에 대한 reponses**
      - **$w_{(i,j)}^{s}$ , $w_{(i,j)}^{t}$** : 위에서 구한 weights를 의미

#### 2.5 Feature Interaction Module
- 위의 과정에서 **최종적으로 선택된 cross-modal prototype vectors**은 **class-related, cross-modal patterns을 모두 포함**하고 있다.
- **Feature Interaction module**을 통해서 **informative patterns을 single-modal features에 포함**시키게 된다.
- **Process**
![image](https://user-images.githubusercontent.com/97836929/220925700-96ea325b-b5de-40c9-bc12-a7a695166c6e.png)
   1. single-modal features와 이와 관련된 associated responses를 concatenate 진행
   2. 다음으로 **single-modal features과 cross-modal prototype vectors를 융합**하기 위해 **Linear layer**을 거친다. 
- **Feature Interaction Module의 output**은 report를 생성하기 위해 **Transformer module의 input으로 사용**된다.

### 3. Reports Generation with Transformer
- 최종적으로 report를 생성하기 위해 encoder-decoder 구조의 transformer 사용
-  **Process**
   - intermediate representations을 얻기 위해 responsed visual features $l^s$이 Encoder의 input으로 들어간다. 
   - Decoder의 입력으로는 fused textual representation sequence $l^t$ = { $l^t_1, l^t_2,..., l^t_{T-1}$ }과 intermediate representations이 들어가게 된다.
![image](https://user-images.githubusercontent.com/97836929/220935065-73ab0518-a382-47bc-9b43-1764840dc05e.png)

## Improved Multi-Label Contrastive Loss 
- 적어도 하나의 공통 레이블(레이블 0 제외)이 있는 sample을 positive pair로 보고, contrastive loss를 multi-label scenario로 수정
- 만약 두 샘플이 공통 label을 공유하지 않으면 negative pair을 형성하게 된다.
- **vanilla contrastive loss**는 **single-label prototype learning을 위해 design**되어있다. 하지만 본 논문에서 진행하는 task의 training sample의 경우 **multiple categories이 존재**하기 때문에 수정    
   - visual responses $r^s = {r_1^s, r_2^s,.., r_{N^s-1}^s, r_{N^s}^s}$, textual responses $r^t = {r_1^t, r_2^t,.., r_{N^t-1}^t, r_{N^t}^t}$에 대한 loss
![image](https://user-images.githubusercontent.com/97836929/221346435-5f2f6633-7cdb-4d45-bdb6-d098fa9d2cc6.png)
   - $\sigma(), Sim()$ 은 각각 image patch responses에 대한 $L_2$ normalization과 cosine similarity function을 의미
   - 기존의 contrastive loss와의 차이점으로는 위의 식에서 $\theta^{()}$으로 positive similarity를 maximize하는 term이 label difference term으로 변경된 점이다. 
      - 이를 통해 positive-pairs에 대해서 dissimilarity를 고려하여 multi-label에 대해서 좀 더 적합하게 된다. 
![image](https://user-images.githubusercontent.com/97836929/221347438-14ca2a51-c607-4656-9466-94b5ee5d96dc.png)
      - $h_d$ : 두 개의 training samples 사이의 **다른 labels의 수**
      $h_t$ : 두 개의 training samples에 대한 **전체 labels의 수**
- 위에서 설명한 improved contrastive loss는 textual responses에 대해서도 동일하게 진행 : $L^t_{icn}$

### Objective Function
![image](https://user-images.githubusercontent.com/97836929/221347688-6d375b61-22bc-4d79-ba02-5ba662db0f50.png)
- 위에서 설명한 improved multi-label contrastive loss와 cross-entropy loss 함께 사용
- predicted report sequence { $p_i$ }, associated ground truth report { $w_i$ }

# Experiment & Result

## medical report generation benchmarks
- Datasets : IU-Xray, MIMIC-CXR datasets
- natural language processing evaluation metrics : BLEU{1-4}, ROUGE-L, METEOR, CIDEr

### Comparisons with Previous Studies 
![image](https://user-images.githubusercontent.com/97836929/221335378-b995a133-63c8-41f0-abc1-4c24b7fee09b.png)
- **IU-Xray dataset**에 대해서 **XPRONET은 이전 SOTA model인 R2GenCMN보다 높은 성능**을 보여준다.
- **MIMIC-CXR benchmark**에서는 **BLEU-1을 제외한 evaluation matric에서는 XPRONET이 더욱 좋은 성능**을 보여주었지만, BLEU-1에서는 R2GenCMN이 조금더 앞선다.
- IU-Ray에서 좀 더 좋은 성능을 보여주는데, 그 이유 중 하나로 **MIMIC-CXR의 경우 IU-Ray보다 50배 이상 많은 sample의 수를 가지는 data size의 차이**가 있다.

### Ablation Analysis
![image](https://user-images.githubusercontent.com/97836929/221335456-27a89bef-528e-4ddb-bff2-880104fbdc0b.png)
- **case 1 : XPRONET**
   **case 2 : XPRONET w/o CMPNet** (Cross-Modal Prototype(CMP) 없이 visual extractor(ResNet-101)과 the encoder-decoder(Transformer)로만 형성)
   **case 3 : XPRONET w/o PI** (XPRONET without the cross-modal Prototype Initialization(PI) →  the cross-modal prototype matrix is randomly initialized.)
   **case 4 : XPRONET w/o IMLCS** (XPRONET  without  the  improved  multi-label contrastive loss(IMLCS).)
- 위의 결과를 통해 **initializing the cross-modal prototype, cross-modal prototype learning setting, improved multi-label contrastive loss(IMLCS)의 중요성**을 보여준다. 

# Insight

(작성중)

### Paper URL

https://arxiv.org/pdf/2207.04818.pdf
